### ISSUE TITLE
Rule idea: `filename-pattern`

### ISSUE THREAD

**Issue by evilebottnawi** [stylelint/stylelint#1485]
We can get filename from `source.input`
```js
{
source: 
   { input: 
      Input {
        css: 'a {\n  color: #zzzzzz;\n}\n',
        file: '/home/sheo13666q/IdeaProjects/stylelint/src/__tests__/fixtures/invalid-hex.css' },
     start: { line: 1, column: 1 } }
}
```

**Comment by jeddy3** [comment]
This sounds like rule for a file system linter. (Now there's an interesting side project for someone...)

You could have rules like:

- `extension-blacklist/whitelist: []`
- `filename-pattern: ""`
- `file-type-path-blacklist/whitelist: {}`

Say you only want hyphenated lower case css files, only allow JavaScript, CSS, SVGs and Jpegs (disallowing PNG etc...), and only allow images within the "components" folder (and nowhere else):

```js
// .filesystemlinterrc
{
  "rules": {
    "extension-whitelist": ["css", "js", "png", "svg"],
    "filename-pattern": "/[\-a-z0-9]/"
    "file-type-path-whitelist": {
      "png": ["/components/**/"]
    }
  }
}
```

Anyhow, I think we can close this issue here in stylelint.

**Comment by evilebottnawi** [comment]
@jeddy3 i think be good create this as stylelint plugin, what do you think?

**Comment by jeddy3** [comment]
I think it belongs in a filesystem linter as it has nothing to do with linting CSS code.

Like I said though, it could be interesting new project for someone and not one I've seen done before.

**Comment by evilebottnawi** [comment]
@jeddy3 yep, you are right, be good have this linter separately, thanks!

**Comment by ntwb** [comment]
As an aside [eslint-plugin-xo](https://github.com/sindresorhus/eslint-plugin-xo/blob/master/rules/filename-case.js) includes a `filename-case` rule for JS filenames, options are: `camelCase`, `snakeCase`, `kebabCase`, and `pascalCase`

**Comment by felixfbecker** [comment]
What if you want to ensure that the file name matches a top-level selector, e.g.
File: MyComponent.scss
Content:
```scss
.my-component {
  // all styles
}
```

Bonus: check if there is a corresponding MyComponent.tsx

**Comment by jeddy3** [comment]
@felixfbecker You can try to create a [plugin](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/plugins.md) to do that.

**Comment by GeorgeTaveras1231** [comment]
In case people find themselves looking for this. I've found this project. Seems legit. Havent tested yet: https://ls-lint.org/1.x/getting-started/installation.html#curl
